### PR TITLE
Aggregate plots with duplicate time stamps

### PR DESCRIPTION
**Pull request by emilk**
### What
If a user logs multiple scalars to an entity path during the same time sequence, the aggregator needs to kick. Now it does.

image
image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/5161/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/5161/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/5161/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!

- [PR Build Summary](https://build.rerun.io/pr/5161)
- [Docs preview](https://rerun.io/preview/c0a24a900fc3ef23e4629fae2ca6c7a87ca579bd/docs) 
- [Examples preview](https://rerun.io/preview/c0a24a900fc3ef23e4629fae2ca6c7a87ca579bd/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_space_view_time_series/src/aggregation.rs [hunk 1]
```diff
@@ -115,9 +118,7 @@ impl MinMaxAggregator {
             let mut acc_max = points[i + j].clone();
             j += 1;
 
-            while j < window_size
-                && i + j < points.len()
-                && are_aggregatable(&points[i], &points[i + j], window_size)
+            while i + j < points.len() && are_aggregatable(&points[i], &points[i + j], window_size)
```
Comment (teh-cmc) [inline crates/re_space_view_time_series/src/aggregation.rs:121]: I'd do that same change in the average aggregator if only for consistency

File: crates/re_space_view_time_series/src/aggregation.rs [hunk 2]
```diff
@@ -147,9 +148,11 @@ impl MinMaxAggregator {
...
             match self {
                 MinMaxAggregator::MinMax => {
-                    aggregated.push(acc_min);
-                    // Don't push the same point twice.
-                    if j > 1 {
+                    if acc_min == acc_max {
+                        // Don't push the same point twice.
+                        aggregated.push(acc_min);
+                    } else {
+                        aggregated.push(acc_min);
```
Comment (teh-cmc) [inline crates/re_space_view_time_series/src/aggregation.rs:155]: Why the change here? Does this actually modify behavior in any way? Comparing `PlotPoint`s in general is shady and I'd rather avoid if if there's no algorithmic reason for it.

File: crates/re_space_view_time_series/src/util.rs [hunk 3]
```diff
@@ -150,7 +150,14 @@ pub fn apply_aggregation(
     // So it can be displayed in the UI by the SpaceViewClass.
     let num_points_before = points.len() as f64;
 
-    let points = if aggregation_duration > 2.0 {
+    // If the user logged multiples scalars per time stamp, we should aggregate them,
+    // no matter what the aggregation duration (=zoom level) is.
+    let multiple_values_per_time_stamp = || points.windows(2).any(|w| w[0].time == w[1].time);
+
+    let should_aggregate = aggregator != TimeSeriesAggregator::Off
+        && (2.0 <= aggregation_duration || multiple_values_per_time_stamp());
```
Comment (teh-cmc) [inline crates/re_space_view_time_series/src/util.rs:158]: I believe we could handle this special case in a much cheaper way if we do it using an accumulator while we are computing the `PlotPoint`s to begin with (no search for duplicates, no allocating unnecessary `PlotPoint`s, etc).

It's fine as a quick fix for the time being but we should open an issue then.

**Review by teh-cmc — APPROVED** [review]

File: crates/re_space_view_time_series/src/aggregation.rs (hunk 1 above)
Comment (emilk) [reply crates/re_space_view_time_series/src/aggregation.rs:121]: Oh yeah, good catch

File: crates/re_space_view_time_series/src/util.rs (hunk 3 above)
Comment (emilk) [reply crates/re_space_view_time_series/src/util.rs:158]: It will be a lot of duplicated code though, for averaging the plot point properties

Comment (teh-cmc) [reply crates/re_space_view_time_series/src/util.rs:158]: > Can't the same argument be made for the whole of the aggregation; that it can be run in a streaming fashion without first allocating all the plot points?

I'm not sure it can off the top of my head, but at the very least it would be much much more tricky because of fractional pixels, inter-timestamp aggregation, edge cases that require peeking your neighbours, etc.

While this specific case is just constrained to a single timestamp with no shenanigans.

Comment (emilk) [reply crates/re_space_view_time_series/src/util.rs:158]: It is also not worth optimizing this right now; the aggregation is dwarfed by the query time:

image
